Test Plan for Quarkus-874: Production support for Windows

Sgitario opened the pull request.

Sgitario commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,36 @@
+   - In Github CI - for Pull Requests and Daily runs
+ - Workshop Test Suite
+   - New Jenkins Jobs
+
```
Do we need to update OpenShift Test Suite as well? As far I see it, the test suite is not OS dependant.

rsvoboda replied:
For OpenShift Test Suite we should run some selected modules like the deployment strategies to ensure some functionality baseline checks.

Sgitario replied:
Good point. Added.

Sgitario commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,36 @@
+## Impact on resources
+- New Windows instances in our Jenkins environments
+  - Windows instance must support Native compilation (GraalVM, VS Build Tools) and Docker (Windows Containers)
+
```
Should we state the existing limitation around testcontainers / Linux containers on Windows? I think this will be eventually addressed, so we don't need to state it here.

rsvoboda replied:
I would mention that, it causes us a lot of troubles. And the available solution doesn't work out of the box / we (Tomas mainly) spent a lot of hours on this topic.

Sgitario replied:
Added as part of the Impact on resources section and I provided also additional context.

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+## Scope of the testing
+The selected Test Suites will be executed on the selected Windows Platforms to ensure functionality is fully accomplished.
+In order to accommodate this in our current test suites, we need to rename existing jobs in Jenkins in order to differentiate them for those running on RHEL and on Windows platforms. 
+
```
Pls mention that some testsuites will need adjustments for correct execution on Windows.

Sgitario replied:
Added

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+   - New Jenkins Jobs
+ - Combinations of Quarkus extensions Test Suite
+   - In Github CI - for Pull Requests and Daily runs
+ - Workshop Test Suite
```
I think this TS is no longer used. @mjurc can you clarify?

mjurc replied:
We still use it as simple test on bare metal

Sgitario replied:
Then I will keep it here.

rsvoboda replied:
Question is if we need such TS on Windows.
For me BeefyTS coverage is enough, Workshop job just for RHEL to ensure functionality.
Or is there anything specific / risky for Windows why Workshop would be needed?

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+   - New Jenkins Jobs
+
+### Platform testing
+Tests are going to be executed against all tested versions of databases, RHEL and OpenShift versions. 
```
This is not true, we will need to be cautious about the expansion of the executed jobs.

Use something like this:
```
JVM mode on is the focus for Windows platform. This RFE does not include native executable (Mandrel) or any extension that requires native libraries (like Kafka streams client).
```

Sgitario replied:
This section was a leftover, thanks for spoting. I've already updated it using your suggestion.

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+
+## Impact on resources
+- New Windows instances in our Jenkins environments
+  - Windows instance must support Native compilation (GraalVM, VS Build Tools) and Docker (Windows Containers)
```
Native compilation (GraalVM, VS Build Tools) is not required by this RFE, I would drop it from this text.
I want that to be available so we have option to try native binaries on Windows eventhough it's not this RFE requirement.

Docker (Windows Containers) is not what we want. We need Docker with Linux containers support.
Our testsuites rely on testcontainers to assure portability between different CI environments and to avoid dependency on many external services.

I miss impact on PSI capacity and execution time of our tests.
 - Windows instances will be executed next to the RHEL instances => more CPU / memory in the lab would be needed to ensure similar execution time as now.
 - We do not have new capacity in the lab => test execution will be prolonged (doubled in the worse case)
 - Windows virtual machines usually require more resources than RHEL ones, expecting `large` or `xlarge` flavor to be used
 - More jobs pulling artifacts from Artifactory instance => we may need to use more resources for Artifactory

Sgitario replied:
Fixed

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+Tests are going to be executed against all tested versions of databases, RHEL and OpenShift versions. 
+Both the JVM and native mode is going to be tested.
+
+## Impact on resources
```
Add "Impact on Jenkins" section to mention need of new jobs definitions and implication on some jobs naming unification

Sgitario replied:
Done

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+In order to accommodate this in our current test suites, we need to rename existing jobs in Jenkins in order to differentiate them for those running on RHEL and on Windows platforms. 
+
+### Selected Windows Platforms
+- Windows Server 2019
```
Please add a note that we will make sure to have access to Windows 10 machine in PSI when needed in case of fix for customer issue bound to that particular environment.

Sgitario replied:
Added

rsvoboda commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+   - In Github CI - for Pull Requests and Daily runs
+   - New Jenkins Jobs
+ - QuickStarts Test Suite
+   - New Jenkins Jobs
```
I'm thinking if we need both QS and Beefy TS jobs for Windows.
I think we could just have Beefy TS.

Sgitario replied:
Ok, removed

Sgitario commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,37 @@
+
+JIRA link: https://issues.redhat.com/browse/QUARKUS-874
+
+The goal is to ensure some level of support and identify existing limitations on Quarkus side for Windows.
```
Done

Sgitario commented:
@rsvoboda I think I've already addressed all your comments, would you mind to take another look?

Sgitario commented on QUARKUS-874.md:
```diff
@@ -0,0 +1,50 @@
+   - Windows virtual machines usually require more resources than RHEL ones, expecting large or xlarge flavor to be used
+   - More CPU / memory in the lab would be needed to ensure similar execution time as now.
+ - Remote Docker machines instances based on RHEL to build container images by Windows jobs
+   - This is required by the Windows jobs for some test suites that require third parties components such us databases and these components are based on Linux images
```
Removed

Sgitario commented:
@rsvoboda I've incorporated all your commit suggestions and dropped the line you requested.
Would you mind to take another look? From my point of view, the only remaining point is if we should cover Workshop TS in Windows or not. I didn't add it for now.

rsvoboda commented:
I plan to give @mjurc one more day, otherwise merging on Friday 2 PM Brno time.
